File: premarket_alert/providers/tradingview.py
"""TradingView-screener als databron.

Gebruikt hetzelfde scanner-endpoint als de Stock Screener op tradingview.com.
Dat endpoint is niet officieel gedocumenteerd: het kan zonder aankondiging
wijzigen, en je hoort het niet vaker te bevragen dan de screener zelf
(een paar keer per minuut is ruim zat). Wil je een gegarandeerd stabiel
contract, gebruik dan PROVIDER=polygon.
"""

from __future__ import annotations

import logging

from ..httpclient import request_json
from ..models import Mover
from .base import Provider

log = logging.getLogger(__name__)

COLUMNS = [
    "name",
    "description",
    "exchange",
    "type",
    "close",
    "premarket_close",
    "premarket_change",
    "premarket_volume",
    "market_cap_basic",
]


class TradingViewProvider(Provider):
    name = "tradingview"

    def build_payload(self) -> dict:
        cfg = self.config
        # Vraag iets onder de drempel op, zodat het resultaat ook bruikbaar is
        # als een lokaal filter (prijs, volume) er nog iets af haalt.
        server_side_floor = max(cfg.threshold_pct - 10.0, 1.0)
        filters: list[dict] = [
            {"left": "premarket_change", "operation": "greater", "right": server_side_floor},
            {"left": "premarket_volume", "operation": "greater", "right": 0},
        ]
        if cfg.min_price > 0:
            filters.append({"left": "premarket_close", "operation": "egreater", "right": cfg.min_price})
        if cfg.max_price > 0:
            filters.append({"left": "premarket_close", "operation": "eless", "right": cfg.max_price})

        return {
            "filter": filters,
            "options": {"lang": "en"},
            "markets": ["america"],
            "symbols": {"query": {"types": []}, "tickers": []},
            "columns": COLUMNS,
            "sort": {"sortBy": "premarket_change", "sortOrder": "desc"},
            "range": [0, max(self.config.max_results * 4, 100)],
        }

    def fetch(self) -> list[Mover]:
        url = self.config.scanner_url
        if "label-product" not in url:
            url += ("&" if "?" in url else "?") + "label-product=screener-stock"

        data = request_json(
            url,
            method="POST",
            payload=self.build_payload(),
            headers={"Origin": "https://www.tradingview.com", "Referer": "https://www.tradingview.com/"},
            timeout=self.config.http_timeout,
        )
        rows = data.get("data") or [] if isinstance(data, dict) else []
        log.info("TradingView-scanner gaf %d rijen terug", len(rows))
        return [m for m in (self._parse_row(row) for row in rows) if m is not None]

    @staticmethod
    def _parse_row(row: dict) -> Mover | None:
        values = row.get("d") or []
        if len(values) < len(COLUMNS):
            return None
        field = dict(zip(COLUMNS, values))

        change_pct = field.get("premarket_change")
        premarket_price = field.get("premarket_close")
        prev_close = field.get("close")
        if change_pct is None or premarket_price is None or not prev_close:
            return None

        ticker = row.get("s", "")
        exchange = field.get("exchange") or (ticker.split(":")[0] if ":" in ticker else "")

        return Mover(
            symbol=field.get("name") or ticker.split(":")[-1],
            exchange=exchange,
            name=field.get("description") or "",
            price=float(premarket_price),
            prev_close=float(prev_close),
            change_pct=float(change_pct),
            volume=int(field.get("premarket_volume") or 0),
            market_cap=field.get("market_cap_basic"),
            instrument_type=(field.get("type") or "stock").lower(),
        )

File: worker/src/providers/tradingview.ts
/**
 * TradingView-screener als databron.
 *
 * Gebruikt hetzelfde scanner-endpoint als de Stock Screener op tradingview.com.
 * Dat endpoint is niet officieel gedocumenteerd: het kan zonder aankondiging
 * wijzigen, en je hoort het niet vaker te bevragen dan de screener zelf.
 * Wil je een gegarandeerd stabiel contract, gebruik dan PROVIDER=polygon.
 */

import type { Config } from "../config.ts";
import { requestJson, type RequestOptions } from "../http.ts";
import type { Mover } from "../models.ts";

export const COLUMNS = [
  "name",
  "description",
  "exchange",
  "type",
  "close",
  "premarket_close",
  "premarket_change",
  "premarket_volume",
  "market_cap_basic",
] as const;

export function buildPayload(cfg: Config): Record<string, unknown> {
  // Vraag iets onder de drempel op, zodat het resultaat ook bruikbaar is
  // als een lokaal filter (prijs, volume) er nog iets af haalt.
  const serverSideFloor = Math.max(cfg.thresholdPct - 10, 1);
  const filter: Record<string, unknown>[] = [
    { left: "premarket_change", operation: "greater", right: serverSideFloor },
    { left: "premarket_volume", operation: "greater", right: 0 },
  ];
  if (cfg.minPrice > 0) filter.push({ left: "premarket_close", operation: "egreater", right: cfg.minPrice });
  if (cfg.maxPrice > 0) filter.push({ left: "premarket_close", operation: "eless", right: cfg.maxPrice });

  return {
    filter,
    options: { lang: "en" },
    markets: ["america"],
    symbols: { query: { types: [] }, tickers: [] },
    columns: COLUMNS,
    sort: { sortBy: "premarket_change", sortOrder: "desc" },
    range: [0, Math.max(cfg.maxResults * 4, 100)],
  };
}

export function parseRow(row: unknown): Mover | null {
  if (!row || typeof row !== "object") return null;
  const { s, d } = row as { s?: unknown; d?: unknown };
  if (!Array.isArray(d) || d.length < COLUMNS.length) return null;
  const field = Object.fromEntries(COLUMNS.map((name, i) => [name, d[i]])) as Record<string, any>;

  const changePct = field.premarket_change;
  const premarketPrice = field.premarket_close;
  const prevClose = field.close;
  if (changePct == null || premarketPrice == null || !prevClose) return null;

  const ticker = typeof s === "string" ? s : "";
  const exchange = field.exchange || (ticker.includes(":") ? ticker.split(":")[0] : "");

  return {
    symbol: field.name || ticker.split(":").at(-1) || "",
    exchange,
    name: field.description || "",
    price: Number(premarketPrice),
    prevClose: Number(prevClose),
    changePct: Number(changePct),
    volume: Math.trunc(Number(field.premarket_volume) || 0),
    marketCap: field.market_cap_basic ?? null,
    instrumentType: String(field.type || "stock").toLowerCase(),
  };
}

export async function fetchTradingView(cfg: Config, opts: RequestOptions = {}): Promise<Mover[]> {
  let url = cfg.scannerUrl;
  if (!url.includes("label-product")) url += (url.includes("?") ? "&" : "?") + "label-product=screener-stock";

  const data = await requestJson(url, {
    ...opts,
    method: "POST",
    payload: buildPayload(cfg),
    headers: { Origin: "https://www.tradingview.com", Referer: "https://www.tradingview.com/" },
    timeoutMs: cfg.httpTimeoutMs,
  });
  const rows = data && typeof data === "object" ? ((data as { data?: unknown }).data ?? []) : [];
  return (Array.isArray(rows) ? rows : []).map(parseRow).filter((m): m is Mover => m !== null);
}
